samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java

```diff
@@ -0,0 +1,152 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote.couchbase;
+
+import com.couchbase.client.deps.io.netty.buffer.ByteBuf;
+import com.couchbase.client.deps.io.netty.util.ReferenceCountUtil;
+import com.couchbase.client.java.document.BinaryDocument;
+import com.couchbase.client.java.document.Document;
+import com.couchbase.client.java.document.JsonDocument;
+import com.couchbase.client.java.document.json.JsonObject;
+import com.google.common.base.Preconditions;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.apache.samza.SamzaException;
+import org.apache.samza.context.Context;
+import org.apache.samza.table.remote.TableReadFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Single;
+import rx.SingleSubscriber;
+
+
+/**
+ * TableReadFunction implementation for reading from Couchbase. The value type can be either {@link JsonObject} or
+ * any other Object. If the value type is JsonObject, data in JSON format will be read from Couchbase directly as
+ * JsonObject. Otherwise, a {@link org.apache.samza.serializers.Serde} needs to be provided to serialize and
+ * deserialize the value object.
+ * @param <V> Type of values to read from Couchbase
+ */
+public class CouchbaseTableReadFunction<V> extends BaseCouchbaseTableFunction<V>
+    implements TableReadFunction<String, V> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CouchbaseTableReadFunction.class);
+
+  protected final Class<? extends Document<?>> documentType;
+
+  /**
+   * Construct an instance of {@link CouchbaseTableReadFunction}.
+   * @param bucketName Name of the couchbase bucket
+   * @param clusterNodes Some Hosts of the Couchbase cluster. Recommended to provide more than one nodes so that if
+   *                     the first node could not be connected, other nodes can be tried.
+   * @param valueClass Type of values
+   */
+  public CouchbaseTableReadFunction(String bucketName, List<String> clusterNodes, Class<V> valueClass) {
+    super(bucketName, clusterNodes, valueClass);
+    documentType = JsonObject.class.isAssignableFrom(valueClass) ? JsonDocument.class : BinaryDocument.class;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void init(Context context) {
+    super.init(context);
+    LOGGER.info(String.format("Read function for bucket %s initialized successfully", bucketName));
```

Comment:
can be simplified into:
LOGGER.info("Read function for bucket {} initialized successfully", bucketName);

Reply:
Sounds good.